numba/tests/test_num_threads.py

```diff
@@ -0,0 +1,433 @@
+
+        # Test set_num_threads outside a jitted function
+        set_num_threads(2)
+
+        @njit(parallel=True)
+        def test_func():
+            x = 5
+            buf = np.empty((x,))
+            for i in prange(x):
+                buf[i] = get_num_threads()
+            return buf
+
+        @guvectorize(['void(int64[:])'],
+                     '(n)',
+                     nopython=True,
+                     target='parallel')
+        def test_gufunc(x):
+            x[:] = get_num_threads()
+
+
+        out = test_func()
+        np.testing.assert_equal(out, 2)
+
+        x = np.zeros((5000000,), dtype=np.int64)
+        test_gufunc(x)
+        np.testing.assert_equal(x, 2)
+
+    @skip_parfors_unsupported
+    @unittest.skipIf(config.NUMBA_NUM_THREADS < 2, "Not enough CPU cores")
+    def test_set_num_threads_inside_jit(self):
+        # Test set_num_threads inside a jitted function
+        @njit(parallel=True)
+        def test_func(nthreads):
+            x = 5
+            buf = np.empty((x,))
+            set_num_threads(nthreads)
+            for i in prange(x):
+                buf[i] = get_num_threads()
+            return buf
+
+        mask = 2
+        out = test_func(mask)
+        np.testing.assert_equal(out, mask)
+
+    @skip_parfors_unsupported
+    @unittest.skipIf(config.NUMBA_NUM_THREADS < 2, "Not enough CPU cores")
+    def test_set_num_threads_inside_guvectorize(self):
+        # Test set_num_threads inside a jitted guvectorize function
+        @guvectorize(['void(int64[:])'],
+                     '(n)',
+                     nopython=True,
+                     target='parallel')
+        def test_func(x):
+            set_num_threads(x[0])
+            x[:] = get_num_threads()
+
+        x = np.zeros((5000000,), dtype=np.int64)
+        mask = 2
+        x[0] = mask
+        test_func(x)
+        np.testing.assert_equal(x, mask)
+
+    @skip_parfors_unsupported
+    @unittest.skipIf(config.NUMBA_NUM_THREADS < 2, "Not enough CPU cores")
+    def test_get_num_threads_truth_outside_jit(self):
+
+        for mask in range(2, min(6, config.NUMBA_NUM_THREADS + 1)):
+            set_num_threads(mask)
+
+            # a lot of work, hopefully will trigger "mask" count of threads to
+            # join the parallel region (for those backends with dynamic threads)
+            @njit(parallel=True)
+            def test_func():
+                x = 5000000
+                buf = np.empty((x,))
+                for i in prange(x):
+                    buf[i] = get_thread_num()
+                return len(np.unique(buf)), get_num_threads()
+
+            out = test_func()
+            self.assertEqual(out, (mask, mask))
+
+            @guvectorize(['void(int64[:], int64[:])'],
+                         '(n), (m)',
+                         nopython=True,
+                         target='parallel')
+            def test_gufunc(x, out):
+                x[:] = get_thread_num() # XXX: Doesn't actually work
```

Comment:
Fixing https://github.com/numba/numba/pull/4615/files#r356589034 should make it compile.